napari/_vispy/layers/labels.py

```diff
@@ -120,60 +161,220 @@ def __init__(
         super().__init__(colors, controls=None, interpolation='zero')
-        self.glsl_map = direct_lookup_shader.replace(
-            '$use_selection', str(use_selection).lower()
-        ).replace('$selection', str(selection))
+        self.glsl_map = (
+            direct_lookup_shader.replace(
+                "$use_selection", str(use_selection).lower()
+            )
+            .replace("$selection", str(selection))
+            .replace("$collision", str(collision).lower())
+            .replace("$color_count", str(color_count))
+        )
 
 
-def idx_to_2D(idx, shape):
+def idx_to_2d(idx, shape):
     """
     From a 1D index generate a 2D index that fits the given shape.
 
     The 2D index will wrap around line by line and back to the beginning.
     """
-    return (idx // shape[1]) % shape[0], (idx % shape[1])
+    return int((idx // shape[1]) % shape[0]), int(idx % shape[1])
 
 
-def hash2d_get(key, keys, values, empty_val=0):
+def hash2d_get(key, keys, col_key) -> Tuple[Optional[Tuple[int, int]], bool]:
     """
-    Given a key, retrieve its location in the keys table.
+    Given a key, retrieve its location in the keys table and information if
+    it is in the main table or collision table.
+    If the key is not found, return None.
     """
-    pos = idx_to_2D(key, keys.shape)
-    initial_key = key
-    while keys[pos] != key and keys[pos] != empty_val:
-        if key - initial_key > keys.size:
-            raise KeyError('label does not exist')
-        key += 1
-        pos = idx_to_2D(key, keys.shape)
-    return pos if keys[pos] == key else None
+    pos = idx_to_2d(key, keys.shape)
+    if keys[pos] == key:
+        return pos, True
+    x = key % col_key.shape[0]
+    for i in range(col_key.shape[1]):
+        if col_key[x, i] == key:
+            return (x, i), False
+    return None, False
 
 
-def hash2d_set(key, value, keys, values, empty_val=0):
+def hash2d_set(key: float, value, keys, values, empty_val=0) -> bool:
     """
     Set a value in the 2d hashmap, wrapping around to avoid collision.
     """
-    if key is None:
-        return
-    pos = idx_to_2D(key, keys.shape)
-    initial_key = key
-    while keys[pos] != empty_val:
-        if key - initial_key > keys.size:
-            raise OverflowError('too many labels')
-        key += 1
-        pos = idx_to_2D(key, keys.shape)
-    keys[pos] = key
-    values[pos] = value
-
-
-def build_textures_from_dict(color_dict, empty_val=0, shape=(1000, 1000)):
+    if key is None or isnan(key):
+        return False
+    pos = idx_to_2d(key, keys.shape)
+    if keys[pos] == key:
+        # casting to float32 and collision with a close key for values above 2**23
+        return False
+    if keys[pos] == empty_val:
+        keys[pos] = key
+        values[pos] = value
+        return False
+
+    return True
+
+
+def _get_shape_from_keys(
+    keys: np.ndarray, fst_dim: int, snd_dim: int
```

Comment:
Would be best to rename variables here to be more explicit, like `second` instead of `snd`. What does `crd` stand for, actually?

Also, some comments or a docstring in this function would be great, I don't understand how those numbers and calculations are optimizing for low collisions, and it would be hard to maintain without asking you...

Reply:
updated, could you check?